vultr/resource_vultr_iso_private.go

```diff
@@ -107,7 +109,57 @@ func resourceVultrIsoDelete(ctx context.Context, d *schema.ResourceData, meta in
 	log.Printf("[INFO] Deleting iso : %s", d.Id())
 
 	if err := client.ISO.Delete(ctx, d.Id()); err != nil {
-		return diag.Errorf("error destroying ISO %s : %v", d.Id(), err)
+		// decode the error
+		var attachedErr struct {
+			Error  string `json:"error"`
+			Status int    `json:"status"`
+		}
+
+		if unmarshalError := json.Unmarshal([]byte(err.Error()), &attachedErr); unmarshalError != nil {
+			return diag.Errorf("error deleting ISO %s: parsing error %v in deleting ISO %s : %v", d.Id(), err.Error(), unmarshalError)
+		}
+
+		if !strings.Contains(attachedErr.Error, "is still attached to") {
+			return diag.Errorf("error deleting ISO %s: delete ISO error not related to attachment: delete error %+v", d.Id(), attachedErr)
+		}
+
+		parts := strings.Split(attachedErr.Error, " ")
+		ip := parts[len(parts)-1]
+		if parsedIP := net.ParseIP(ip); parsedIP == nil {
+			return diag.Errorf("error deleting ISO %s : failed to parse IP to which ISO is attached: %s", d.Id(), ip)
+		}
+
+		for {
+			// default is 100 instances
+			var options govultr.ListOptions
+			instances, meta, err := client.Instance.List(ctx, &options)
+			if err != nil {
+				return diag.Errorf("error deleting ISO %s : failed to list instances for detaching ISO", d.Id(), err)
+			}
+
+			// check for the instance with this IP, return on failure or discovery
+			for _, instance := range instances {
+				if instance.MainIP == ip {
+					if err := client.Instance.DetachISO(ctx, instance.ID); err != nil {
+						return diag.Errorf("error deleting ISO %s : failed to detach from instances %s : %s", d.Id(), instance.ID, err)
+					}
+					_, err := waitForIsoDetached(ctx, instance.ID, "ready", []string{"isomounted"}, "status", meta)
+					if err != nil {
+						return diag.Errorf("error deleting ISO %s: failed to wait for ISO to detach from instance %s: %s", d.Id(), instance.ID, err)
+					}
+					if err = client.ISO.Delete(ctx, d.Id()); err != nil {
+						return diag.Errorf("error deleting ISO %s: failed to delete ISO: %s", d.Id(), err)
+					}
+				}
+				options.Cursor = meta.Links.Next
+			}
+
+			// no more instances to check
+			if options.Cursor == "" {
```

Comment:
I would keep this in line with how we are doing the paging loops in other places by keeping it as 


```
	if meta.Links.Next == "" {
				break
			}
	options.Cursor = meta.Links.Next
```

and moving 

https://github.com/vultr/terraform-provider-vultr/pull/131/files#diff-9440d6827d830f384c9118b4fc75b2aea56b1c826f826fc780dbef8c208831f5R134

above the initial for loop as `options := &govultr.ListOptions{}`

Reply:
Oh, duh. 1 sec!

Reply:
Addressed in https://github.com/vultr/terraform-provider-vultr/pull/131/commits/3421f2452245d4b21281ac740bd8f7f5402c40d8 .